### PR TITLE
[CVP][SCCP] Add support for `uitofp nneg`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[CVP] Add tests for adding `nneg` flag to `uitofp` and converting `sitofp` -> `uitofp nneg`; NFC**
- **[CVP] Convert `sitofp` -> `uitofp nneg` and add `nneg` flag to `uitofp`**
- **[SCCP] Add `nneg` flag to `uitofp` if its operand is non-negative**

### CODE REVIEW

**Comment by goldsteinn** [comment]
I looked into adding support in `SCEV` but it didn't seem worth it to
add complete support for `uitofp`/`sitofp` just for this. Any opinions
on if I should revisit / any other passes that I should add support
for? Mostly I just copied where we do the same for `zext`.

**Comment by goldsteinn** [comment]
rebased

**Comment by nikic** [comment]
Can you please split this into separate patches? Or at least split out the InstCombine part of it?

**Comment by goldsteinn** [comment]
rebase/ping

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 1]
```diff
@@ -1075,20 +1076,50 @@ static bool processSExt(SExtInst *SDI, LazyValueInfo *LVI) {
   return true;
 }
 
-static bool processZExt(ZExtInst *ZExt, LazyValueInfo *LVI) {
-  if (ZExt->getType()->isVectorTy())
+static bool processPossibleNonNeg(Instruction *I, LazyValueInfo *LVI) {
+  assert(isa<PossiblyNonNegInst>(I) && "Requires PossiblyNonNeg instruction");
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1080]: Can you use `PossiblyNonNegInst *I` as param type instead? Or is the hierarchy not right for that currently?

File: llvm/test/Transforms/CorrelatedValuePropagation/sitofp.ll [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+for.body:  ; preds = %for.cond
+  %ext.wide = sitofp i32 %a to float
+  call void @use.f32(float %ext.wide)
+  %ext = fptosi float %ext.wide to i32
+  br label %for.cond
+
+for.end:  ; preds = %for.cond
+  ret void
+}
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/sitofp.ll:38]: These tests look unnecessarily complicated -- can we just assert that %a is non-negative? Or at least drop the loop around it?

File: llvm/test/Transforms/CorrelatedValuePropagation/sitofp.ll [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+;; Non looping test case.
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/sitofp.ll:109]: Okay, I see you have a test without loop here -- but what's the point of the looping tests?

Comment (goldsteinn) [reply llvm/test/Transforms/CorrelatedValuePropagation/sitofp.ll:109]: The the tests are really just copy-paste of what we have for `zext` + `nneg`, didn't put much thought beyond that.

Comment (nikic) [reply llvm/test/Transforms/CorrelatedValuePropagation/sitofp.ll:109]: Then please reduce the tests to the essentials... You only need the tests below here and can drop the ones above.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**